### PR TITLE
Introduce MethodTypeKind to quotes reflection API

### LINKED ISSUE #18477: Add contextual method types to Quotes API
Hmm, interesting... It looks like the compiler internally uses `ContextualMethodType` rather than just `MethodType` to express a method with `(using ...)` parameters clause but currently there seems no way to create contextual methods from the quotes' reflect API. @nicolasstucki are there any plans to add this functionality in the nearest future?

_Originally posted by @prolativ in https://github.com/lampepfl/dotty/discussions/14056#discussioncomment-6845567_

### PR DESCRIPTION
**Pull request by jchyb**
It allows to create Contextual and Implicit MethodTypes. MethodTypeKind abstracts away the `MethodTypeCompanion` implementation into a simple enum style choice for a newly added MethodType apply method. The MethodType unapply is kept as it was for source compatibility, instead users are encouraged to use `isImplicit` and `isContextual` methods.

Based on #18499 
Fixes #18477

### CODE REVIEW

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -2223,6 +2232,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       extension (self: MethodType)
         def isErased: Boolean = false
         def isImplicit: Boolean = self.isImplicitMethod
+        def isContextual: Boolean = self.isContextualMethod
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2235]: We should also have a method `def methodTypeKind: MethodTypeKind`.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -2204,6 +2204,13 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         case _ => None
     end MethodOrPolyTypeTest
 
+    type MethodTypeKind = dotc.core.Types.MethodTypeCompanion
+
+    object MethodTypeKind extends MethodTypeKindModule:
+      def Plain: MethodTypeKind = Types.MethodType
+      def Contextual: MethodTypeKind = Types.ContextualMethodType
+      def Implicit: MethodTypeKind = Types.ImplicitMethodType
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2212]: Maybe these should be `val`s. This way we could do

```scala
mt.methodTypeKind match 
  case Plain => ...
  case Contextual => ...
  case Implicit => ...
```

**Review by nicolasstucki — COMMENTED** [review]
Otherwise LGTM.

**Review by nicolasstucki — APPROVED** [review]

File: library/src/scala/quoted/Quotes.scala [hunk 3]
```diff
@@ -3234,6 +3236,22 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
     /** `TypeTest` that allows testing at runtime in a pattern match if a `TypeRepr` is a `MethodOrPoly` */
     given MethodOrPolyTypeTest: TypeTest[TypeRepr, MethodOrPoly]
 
+    /** Type which decides on the kind of parameter list represented by `MethodType`. */
+    type MethodTypeKind
```
Comment (sjrd) [inline library/src/scala/quoted/Quotes.scala:3240]: Have you considered making this a concrete, static `enum` instead?

Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3240]: In this case, it could work.

**Comment by Gedochao** [comment]
This has been decided to be included in the 3.5.0 release.

**Comment by jchyb** [comment]
I have changed the implementation to use enums. I guess this is better - users get nicer pattern matching and the MethodTypeCompanion is also better hidden (printlns on MethodTypeKind will be less confusing)

**Review by sjrd — APPROVED** [review]

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**